Set service account scopes on cluster resume

rtitle opened the pull request:
See https://github.com/DataBiosphere/leonardo/issues/495

When we create a cluster, we set a service account and scopes which go into VM metadata. For some reason, if the instance is stopped/started, the service account persists but the scopes do not. This causes errors when using libraries which require certain scopes (e.g. FISS, AoU client) after pause/resume.

This makes Leo re-set the service account and scopes on the VM at cluster resume time. I verified manually that this works, and also added automation tests. 

Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get two thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

rtitle committed:
FISS calls return 401 after cluster pause/resume

rtitle committed:
Add unit test

rtitle committed:
Add automation test

rtitle commented on automation/project/Dependencies.scala:
```diff
@@ -45,6 +45,11 @@ object Dependencies {
       excludeApacheHttpClient,
       excludeGoogleJsr305,
       excludeJacksonCore),
+    "com.google.apis" % "google-api-services-sourcerepo" % s"v1-rev21-1.23.0" excludeAll (
```
Just added for the `CloudSourceRepositoriesScopes` class.

aweng98 replied:
can you upgrade workbench-service-test version to the latest `val serviceTestV = "0.12-1a00943"`? Credential scopes is newly required parameter in `UserAuthToken`. I need `createDownloadDirectory()` in new version of workbench-service-lib.

rtitle replied:
I updated to the latest `workbench-service-test`.

rtitle commented on src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/google/GoogleDataprocDAO.scala:
```diff
@@ -10,7 +10,14 @@ import org.broadinstitute.dsde.workbench.model.google._
+                    clusterServiceAccount: Option[WorkbenchEmail],
+                    credentialsFileName: Option[String],
+                    stagingBucket: GcsBucketName,
+                    serviceAccountScopes: Seq[String]): Future[Operation]
```
Added `serviceAccountScopes`

rtitle commented on src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala:
```diff
@@ -331,15 +339,7 @@ class LeonardoService(protected val dataprocConfig: DataprocConfig,
 
         // Now stop each instance individually
         _ <- Future.traverse(cluster.nonPreemptibleInstances) { instance =>
-          // Install a startup script on the master node so Jupyter starts back up again once the instance is restarted
```
Moved this installation from cluster stop time to cluster start time. I think it's better to set the cluster metadata lazily, right before it's started. This ensures that recent changes are applied to clusters that have been previously auto-paused.

kyuksel replied:
good call 👍

rtitle replied:
Actually... I'm seeing more frequent test failures, "should pause and resume a cluster with preemptible instances" against this PR. I'm wondering if setting the instance metadata needs to be done at stop time, as it was before. I might move that back, but keep `setServiceAccount` at cluster start time.

kyuksel replied:
Hah...it's curious why that'd cause it

rtitle commented:
jenkins retest

rtitle committed:
Pass kernel into verifyNotebookCredentials()

rtitle commented:
jenkins retest

rtitle commented:
I think tests should pass; need to rebase with https://github.com/DataBiosphere/leonardo/pull/505 and re-run.

rtitle commented:
jenkins retest

rtitle commented:
jenkins retest

rtitle commented:
jenkins retest

rtitle committed:
Should probably set SA+scopes on all instances, not just master

rtitle committed:
Enable auto freeze

rtitle commented:
jenkins retest

kyuksel commented on automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/ClusterMonitoringSpec.scala:
```diff
@@ -30,9 +30,9 @@ class ClusterMonitoringSpec extends FreeSpec with LeonardoTestUtils with Paralle
+            withNewNotebook(cluster, PySpark2) { notebookPage =>
               // should not have notebook credentials because Leo is not configured to use a notebook service account
-              verifyNoNotebookCredentials(notebookPage)
+              verifyNoNotebookCredentials(notebookPage, PySpark2, petEmail)
```
The name `verifyNoNotebookCredentials` sounds misleading to me. Given what it does, shouldn't it be something like `verifyNotebookCredentials`, or something more explanatory?

kyuksel replied:
But I now see that there _is_ also a `verifyNotebookCredentials`, which verifies a subset of things. My suggestion would be to refactor those two methods like so:

```scala
def verifyOauth2Client()
def verifyFiss()
def verifySpark()
def verifyMetadataServer() // preferably something even more specific like verifyScopes()
```
and then if it is necessary, we can have some convenience methods that bundle some of those together as necessary.

rtitle replied:
👍 , agreed that looks much better.

kyuksel commented on src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/google/HttpGoogleComputeDAO.scala:
```diff
@@ -156,6 +156,13 @@ class HttpGoogleComputeDAO(appName: String,
     }
   }
 
+  override def setServiceAccount(instanceKey: InstanceKey, serviceAccountEmail: WorkbenchEmail, serviceAccountScopes: Seq[String]): Future[Unit] = {
```
Nit: Can we stack the input parameters to shorten line length?

kyuksel commented on src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/google/HttpGoogleDataprocDAO.scala:
```diff
@@ -212,7 +207,7 @@ class HttpGoogleDataprocDAO(appName: String,
   }
 
-  private def getClusterConfig(machineConfig: MachineConfig, initScript: GcsPath, clusterServiceAccount: Option[WorkbenchEmail], credentialsFileName: Option[String], stagingBucket: GcsBucketName): DataprocClusterConfig = {
+  private def getClusterConfig(machineConfig: MachineConfig, initScript: GcsPath, clusterServiceAccount: Option[WorkbenchEmail], credentialsFileName: Option[String], stagingBucket: GcsBucketName, serviceAccountScopes: Seq[String]): DataprocClusterConfig = {
```
Nit: Can we stack the input parameters to shorten line length?

kyuksel commented on src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala:
```diff
@@ -100,10 +103,15 @@ class LeonardoService(protected val dataprocConfig: DataprocConfig,
   // The || clause is included because older clusters may not have the run-jupyter.sh script installed,
   // so we need to fall back running `jupyter notebook` directly. See https://github.com/DataBiosphere/leonardo/issues/481.
-  private lazy val masterInstanceStartupScript: immutable.Map[String, String] = {
+  private[service] lazy val masterInstanceStartupScript: immutable.Map[String, String] = {
```
Could remove the `immutable`s since `Map` is already an alias for it.

kyuksel reviewed (commented):
The changes look good overall 👍 Minor comments inline. Also wondering if the scope change warrants a review by the security team?

rtitle commented:
Rebased, I'm going to clean up and use the PR to address https://github.com/DataBiosphere/leonardo/issues/542 as well.

rtitle committed:
PR feedback, and added cloud-platform scopes

rtitle committed:
Fix test assertion

rtitle commented:
jenkins retest

rtitle committed:
Remove cloud platform scopes, it's more complicated

rtitle committed:
Move setting instance metadata to instance stop time as it was before -- it didn't seem to be taking effect when set at instance start time.

rtitle commented:
Rebased this PR and fixed an [outstanding issue](https://github.com/DataBiosphere/leonardo/pull/503/commits/1abd2266373c56b51407be7932072f09074dbabc) where instance metadata was being set in the wrong place. I think this should be in good shape now -- need to verify automation tests.

rtitle commented:
gpalloc issues -- jenkins retest

kyuksel commented:
The following paragraph on the [online doc](
https://software.broadinstitute.org/firecloud/documentation/article?id=11603) will become out of date with this PR:

>FireCloud will soon be able to automatically pause idle clusters and sync your notebooks back to your
workspace. For now though, please be mindful of both running costs and the need to save your data. We don't want anyone to suffer from unhappy surprises.

rtitle commented:
jenkins retest

rtitle commented:
I am closing this for now because it's not quite ready.
